### PR TITLE
server: allows compute offering with or without constraints

### PR DESCRIPTION
**Pull request by shwstppr**
## Description
**Problem**: Custom compute offering does not allow setting min and max values for CPU and VRAM for custom VMs. 

**Root Cause**: Custom compute offerings cannot be created with a given range of CPU number and memory instead it allows only fixed values. 

**Solution**: createServiceOffering API has been modified to allow setting a defined range for CPU number and memory. Also, UI form for compute offering creation is provided with a new field named 'compute offering type’ with values - Fixed, Custom Constrained, Custom Constrained. It will allow the creation of compute offerings either with a fixed CPU speed and memory for fixed compute offering, or with a range of CPU number and memory for custom constrained compute offering or without predefined CPU number, CPU speed and memory for custom unconstrained compute offering. 

To allow the user to set CPU number, CPU speed and memory during VM deployment, UI form for VM deployment has been modified to provide controls to change these values. These controls are depicted in screenshots below for custom constrained and custom unconstrained compute offering types.

Sample API calls using cmk to create a constrained service offering and deploying a VM using it,

`create serviceoffering name=Constrained displaytext=Constrained customized=true mincpunumber=2 maxcpunumber=4 cpuspeed=400 minmemory=256 maxmemory=1024`

`deploy virtualmachine displayname=ConstrainedVM serviceofferingid=60f3e500-6559-40b2-9a61-2192891c2bd6 templateid=8e0f4a3e-601b-11e9-9df4-a0afbd4a2d60 zoneid=9612a0c6-ed28-4fae-9a48-6eb207af29e3 details[0].cpuNumber=3 details[0].memory=800`

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
Create compute offering form
Screenshot from 2019-03-12 17-32-13
Screenshot from 2019-03-12 17-32-17
Screenshot from 2019-03-12 17-32-22

VM deployment form
Screenshot from 2019-03-12 17-43-52
Screenshot from 2019-03-12 17-44-01

### CODE REVIEW

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: ui/l10n/en.js [hunk 1]
```diff
@@ -554,6 +554,10 @@ var dictionary = {
 "label.compute":"Compute",
 "label.compute.and.storage":"Compute and Storage",
 "label.compute.offering":"Compute offering",
+"label.compute.offering.type":"Compute offering type",
+"label.compute.offering.custom.constrained":"Custom Constrained",
+"label.compute.offering.custom.unconstrained":"Custom Unsconstrained",
```
Comment (PaulAngus) [inline ui/l10n/en.js]:
```suggestion
"label.compute.offering.custom.unconstrained":"Custom Unconstrained",
```

Comment (borisstoyanov) [reply ui/l10n/en.js]: Good hack. Simply the ui typing field could be grey out but that won't solve the api call. This need a service layer check imho.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by PaulAngus** [comment]
@blueorangutan package

**Comment by PaulAngus** [comment]
retested UI and API.
it is possible (by moving and forwards in the wizard and selecting different constrained custom offerings to confuse the UI into letting you request an offering that doesn't meet the criteria BUT the API then rejects it anyway. so it's a very very niche case, that I'm not concerned about.

LGTM.

**Comment by rhtyd** [comment]
Merging based on reviews from Paul and Bobby, and the smoketests.

**Pull request merged**